src/common/fd-mux.cc

```diff
@@ -0,0 +1,172 @@
+/* -*- mode: c++; c-basic-offset: 2; indent-tabs-mode: nil; -*-
+ * (c) 2018 Henner Zeller <h.zeller@acm.org>
+ *          Leonardo Romor <leonardo.romor@gmail.com>
+ *
+ * This file is part of BeagleG. http://github.com/hzeller/beagleg
+ *
+ * BeagleG is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * BeagleG is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with BeagleG.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "fd-mux.h"
+#include "logging.h"
+
+#include <unistd.h>
+#include <signal.h>
+#include <vector>
+#include <algorithm>
+#include <sys/select.h>
+#include <errno.h>
+#include <string.h>
+
+static volatile sig_atomic_t caught_signal = 0;
+
+static void receive_signal(int signo) {
+  static char msg[] = "Caught signal. Shutting down ASAP.\n";
+  if (!caught_signal) {
+    write(STDERR_FILENO, msg, sizeof(msg));
+  }
+  caught_signal = 1;
+}
+
+static void arm_signal_handler() {
+  caught_signal = 0;
+  struct sigaction sa = {};
+  sa.sa_handler = receive_signal;
+  sa.sa_flags = SA_RESETHAND;  // oneshot, no restart
+  sigaction(SIGTERM, &sa, NULL);  // Regular kill
+  sigaction(SIGINT, &sa, NULL);   // Ctrl-C
+
+  // Other, internal problems that should never happen, but
+  // can trigger multiple times before we gain back control
+  // to shut down as cleanly as possible. These are not one-shot.
+  sa.sa_flags = 0;
+  sigaction(SIGSEGV, &sa, NULL);
+  sigaction(SIGBUS, &sa, NULL);
+  sigaction(SIGFPE, &sa, NULL);
+}
+
+static void disarm_signal_handler() {
+  signal(SIGTERM, SIG_DFL);  // Regular kill
+  signal(SIGINT, SIG_DFL);   // Ctrl-C
+  signal(SIGSEGV, SIG_DFL);
+  signal(SIGBUS, SIG_DFL);
+  signal(SIGFPE, SIG_DFL);
+}
+
+bool FDMultiplexer::RunOnReadable(int fd, const Handler &handler) {
+  return read_handlers_.insert({ fd, handler }).second;
+}
+
+bool FDMultiplexer::RunOnWritable(int fd, const Handler &handler) {
+  return write_handlers_.insert({ fd, handler }).second;
+}
+
+void FDMultiplexer::RunOnIdle(const Handler &handler) {
+  idle_handlers_.push_back(handler);
+}
+
+bool FDMultiplexer::SingleCycle(unsigned int timeout_ms) {
+  fd_set read_fds;
+  fd_set write_fds;
+
+  struct timeval timeout;
+  timeout.tv_sec = timeout_ms / 1000;
+  timeout.tv_usec = (timeout_ms % 1000) * 1000;
+
+  int maxfd = -1;
+  FD_ZERO(&read_fds);
+  FD_ZERO(&write_fds);
+
+  // Readers
+  for (const auto &it : read_handlers_) {
+    if (it.first >= maxfd) maxfd = it.first + 1;
```

Comment:
Here, we are mixing two things which might be harder to understand for the reader.
You are doing `it.first + 1`, and the reaser might ask why not just it.fisrt.
This is because select() wants the highest filedescriptor + 1 later. But the reader does not know that yet, and the variable does not reflect that (after all, it is not called `max_fd_plus_one`).
And this also means that we have this **>=** test, which is a bit confusing.

I suggest that we just collect the hightest fds here, simply with
```
   maxfd = std::max(maxfd, it.first);
```

and then later, in the call to select do
```
  select(maxfd + 1, ....)
```

This helps reading the code.